apps/src/lab2/responseValidators.ts

```diff
@@ -52,3 +80,32 @@ export class ValidationError extends Error {
     Object.setPrototypeOf(this, ValidationError.prototype);
   }
 }
+
+export function setValidatorForAppType(appName: AppName) {
+  if (appName === 'pythonlab') {
+    Lab2Registry.getInstance().setSourceResponseValidator(
+      PythonSourceResponseValidator
+    );
+  } else if (BLOCKLY_LABS.includes(appName)) {
+    // Blockly labs
+    Lab2Registry.getInstance().setSourceResponseValidator(
+      BlocklySourceResponseValidator
+    );
+  } else {
+    // Everything else uses the default validator
+    Lab2Registry.getInstance().setSourceResponseValidator(
+      DefaultSourceResponseValidator
+    );
+  }
+}
+
+function sourceValidatorHelper(
```

Comment:
What do you think about something like this to reduce some of the repeated code?

```
function parseJSON(source: string): any {
  try {
    return JSON.parse(source);
  } catch (e) {
    throw new ValidationError(`Error parsing JSON: ${e}`);
  }
}

function checkRequiredField(object: Record<string, any>, field: string): void {
  if (object[field] === undefined) {
    throw new ValidationError(`Missing required field: ${field}`);
  }
}

function sourceValidatorHelper(
  response: Record<string, unknown>,
  validate: (source: any) => void
): ProjectSources {
  checkRequiredField(response, 'source');
  const parsedSource = parseJSON(response.source as string);
  validate(parsedSource);
  return response as ProjectSources;
}

export const BlocklySourceResponseValidator: ResponseValidator<ProjectSources> = response => {
  return sourceValidatorHelper(response, blocklySource => {
    checkRequiredField(blocklySource, 'blocks');
  });
};

export const PythonSourceResponseValidator: ResponseValidator<ProjectSources> = response => {
  return sourceValidatorHelper(response, pythonSource => {
    checkRequiredField(pythonSource, 'main.py');
  });
};

export const DefaultSourceResponseValidator: ResponseValidator<ProjectSources> = response => {
  checkRequiredField(response, 'source');
  return response;
};
```
Forgive any TypeScript mistakes here!

Reply:
will update!

Reply:
I did a modified version of this. Instead of a `checkRequiredField` method I made a `throwMissingFieldError` method that handles throwing the exception. The typing was different for each version of the method so it would be messy to try to make it work as one method. It will also get more complicated for python lab once we have nested folders.